Access token service doctrine

padmasreegade opened the pull request:
Raising PR to compare. This is not ready for review yet and the code is not fully functional. There are some failing tests that need to be addressed.

padmasreegade committed:
initial commit

padmasreegade committed:
with tests

padmasreegade committed:
test cases

padmasreegade committed:
changes in the interface and implementation

padmasreegade committed:
testcases

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/AccessToken.php:
```diff
@@ -78,15 +79,15 @@ class AccessToken implements AccessTokenEntityInterface
      * User ID.
      *
-     * @var User
+     * @var int
```
Based on the way Doctrine works, this needs to be a User object -- the ORM layer will automatically populate the object based on the relationship annotations below. Code that needs a user ID will have to call $entity->getUser()->getId(). Code that sets the user may need to access to user service if it only has an ID, so it can look up the object that it needs to set.

demiankatz commented on module/VuFind/src/VuFind/Db/Row/PluginManager.php:
```diff
@@ -46,7 +46,7 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
      */
     protected $aliases = [
-        'accesstoken' => AccessToken::class,
+        //'accesstoken' => AccessToken::class,
```
It should be safe to delete these lines instead of just commenting them out.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -83,7 +77,19 @@ public function getByIdAndType(
+            . 'WHERE id = :id '
+            . 'AND type = :type';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters(['id' => $id, 'type' => $type]);
```
Purely optional, but when you have a construction like this where array keys have the same name as the variables being assigned to them, you can shorten it with the `compact` function -- e.g. `['id' => $id, 'type' => $type]` becomes `compact('id', 'type')`.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -83,7 +77,19 @@ public function getByIdAndType(
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters(['id' => $id, 'type' => $type]);
+        $result = $query->getResult();
+        $result = $result ?: $this->createEntity()
```
I think you may need to account for the `$create` argument here -- return the new object if `$create` is true, but return null if it is false.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -108,7 +118,14 @@ public function storeNonce(int $userId, ?string $nonce): void
-        return $this->accessTokenTable->getNonce($userId);
+        $dql = 'SELECT data '
+            . 'FROM ' . $this->getEntityClass(AccessToken::class) . ' at '
+            . 'WHERE at.user_id = :userId';
```
Doctrine uses property names rather than database table rows, so there is no such thing as `user_id`. I believe this should be changed to `user` (assuming that the entity class is also adjusted to use user instead of userId, which I think would be the correct thing to do).

Also note that if you want to turn an integer into a user object for the purposes of making a query, there's a helper function for that: `$user = $this->getDoctrineReference(User::class, $userId);`. (That example assumes that you've included a use statement for `VuFind\Db\Entity\User`).

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -121,6 +138,18 @@ public function getNonce(int $userId): ?string
+        $subQueryBuilder->select('at.id')
+            ->from($this->getEntityClass(AccessTokenEnTityInterface::class), 'at')
+            ->where('at.created < :latestCreated')
+            ->setParameter('lastestCreated', $dateLimit->getTimestamp());
```
There's a typo here -- latestCreated vs. lastestCreated.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/PluginManager.php:
```diff
@@ -46,7 +46,7 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
      */
     protected $aliases = [
-        'accesstoken' => AccessToken::class,
+        //'accesstoken' => AccessToken::class,
```
Similarly, it should be safe to delete the lines here as well.

demiankatz commented on module/VuFind/src/VuFind/OAuth2/Repository/AbstractTokenRepository.php:
```diff
@@ -99,7 +99,7 @@ public function persistNew($token)
         $userIdentifierField = $this->oauth2Config['Server']['userIdentifierField'] ?? 'id';
         $user = $this->userService->getUserByField($userIdentifierField, $userIdentifier);
-        $row->setUser($user);
+        $row->setUser($userIdentifier);
```
This needs to be reverted -- setUser should accept a user object. Also, this revised version of the logic will only work when $userIdentifierField is set to 'id', which is not guaranteed. If we look up users by a different field, then it will break.

demiankatz reviewed (changes requested):
Thanks, @padmasreegade, this is off to a good start, but I spotted a few issues that I think will prevent this from working correctly. I haven't looked closely at the tests yet, since I think it's probably worth fixing the underlying problems first and then seeing how the tests respond. Let me know if you have questions about anything or if I can do more to help, and thanks for all the work on this! I think it's getting close to completion.

padmasreegade committed:
Changed according to comments:
Modified entity interface to align with Doctrine ORM.
Modified services and tests to keep in sync with the updated interface.

padmasreegade committed:
modified tests

padmasreegade committed:
syncing with DB schema

demiankatz committed:
Progress on test; minor fixes.

demiankatz committed:
Fixes.

padmasreegade committed:
fixed test

padmasreegade committed:
added mocks for missing methods

padmasreegade committed:
fixed stan issues

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -83,7 +78,21 @@ public function getByIdAndType(
+        $result = $query->getOneOrNullResult();
+        if ($result === null && $create) {
+            $result = $this->createEntity()
+                           ->setType($type)
```
I think you're missing a `setId()` call here.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -96,7 +105,11 @@ public function getByIdAndType(
+        $dql = 'UPDATE ' . $this->getEntityClass(AccessToken::class) . ' at '
+                . 'SET at.data = :nonce WHERE at.user = :user';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters(['nonce' => $nonce, 'user' => $userId]);
```
Probably worth using `$this->entityManager->getReference(UserEntityInterface::class, $userId)` here -- will increase flexibility if we want to broaden the type of $userId in future.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -121,6 +149,18 @@ public function getNonce(int $userId): ?string
-        return $this->accessTokenTable->deleteExpired($dateLimit->format('Y-m-d H:i:s'), $limit);
+        $subQueryBuilder = $this->entityManager->createQueryBuilder();
+        $subQueryBuilder->select('at.id')
+            ->from($this->getEntityClass(AccessTokenEnTityInterface::class), 'at')
```
There's a minor case-related typo here and in other instances of the interface name -- note the extra capital T in the middle of EnTity.

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/OAuth2/Repository/AbstractTokenRepositoryTestCase.php:
```diff
@@ -172,57 +137,126 @@ protected function getMockUserTable(): User
+     */
+    protected function getPluginManager($setExpectation = false)
+    {
+        $pluginManager = $this->getMockBuilder(
```
You could simplify this to `$pluginManager = $this->createMock(\VuFind\Db\Entity\PluginManager::class);` since you're not using any non-default behaviors of the mock builder.

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/OAuth2/Repository/AbstractTokenRepositoryTestCase.php:
```diff
@@ -232,26 +266,94 @@ protected function createUserRow(array $data): UserRow
         $accessTokenService->expects($this->any())
-            ->method('storeNonce')
-            ->willReturnCallback([$accessTokenTable, 'storeNonce']);
+        ->method('storeNonce')
```
These lines appear to have been accidentally de-indented.

demiankatz reviewed (changes requested):
Thanks, @padmasreegade, this is looking very good. I have just a few very minor suggestions/observations below.

padmasreegade committed:
fixed according to comments

padmasreegade committed:
modified service

padmasreegade committed:
updated according to old functionality

padmasreegade committed:
modified according to old functionality - using userId as id for token type openid_nonce

padmasreegade committed:
stan fixes

padmasreegade committed:
length issues

padmasreegade committed:
modified to persist Accesstoken Entity

padmasreegade committed:
modified according to comments

demiankatz committed:
Merge branch 'doctrine' into access-token-service-doctrine

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -92,7 +93,8 @@ public function getByIdAndType(
         }
-
+        
+        $this->persistEntity($result);
```
I think you should move this inside the `if ($result === null && $create) {` clause. You should only persist the data if you've been asked to create it. Otherwise you may perform unnecessary writes and/or illegally pass null to the persist method.

padmasreegade committed:
Modified to get nonce alone instead of the whole accesstoken object

padmasreegade committed:
code correction

padmasreegade committed:
whitespace and stan issues fixed

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -108,7 +131,12 @@ public function storeNonce(int $userId, ?string $nonce): void
-        return $this->accessTokenTable->getNonce($userId);
+        $type = 'openid_nonce';
+        $token = $this->getByIdAndType((string)$userId, $type, false);
+        if ($token == null) {
```
You can simplify lines 136-139 to a single line using the null-safe operator: `return $token?->getData();`

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -83,7 +78,23 @@ public function getByIdAndType(
         bool $create = true
     ): ?AccessTokenEntityInterface {
-        return $this->accessTokenTable->getByIdAndType($id, $type, $create);
+
```
For style reasons, I think you might want to remove this blank line at the top of the function.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/AccessTokenService.php:
```diff
@@ -96,7 +107,19 @@ public function getByIdAndType(
+        $token = $this->getByIdAndType((string)$userId, $type);
+        $token->setUser($this->entityManager->getReference(User::class, $userId));
+        $token->setData($nonce);
+        $dql = 'UPDATE ' . $this->getEntityClass(AccessToken::class) . ' a '
```
I don't think you need to do an UPDATE query here. Wouldn't it be easier to replace this whole update query with `$this->persistEntity($token);`? I think that would do the same thing. If it's not working as expected, perhaps you also need to flush the entity manager to be sure the database is updated immediately.

demiankatz reviewed (changes requested):
@padmasreegade, all tests are passing in my VM, though I have a few minor suggestions below.

Are you still seeing failures on your end? If so, we should figure out what is different between our environments....

padmasreegade committed:
Fixed according to comments

padmasreegade committed:
Merge remote-tracking branch 'demiankatz/doctrine' into access-token-service-doctrine

demiankatz committed:
Small fix.

demiankatz committed:
Merge branch 'doctrine' into access-token-service-doctrine

demiankatz committed:
Account for multi-part primary key in expiration.

demiankatz reviewed (approved):
Looks good, and all tests passing!
